atom/renderer/api/atom_api_renderer_ipc.cc

```diff
@@ -27,40 +33,91 @@ RenderFrame* GetCurrentRenderFrame() {
   return RenderFrame::FromWebFrame(frame);
 }
 
-void Send(mate::Arguments* args,
-          bool internal,
-          const std::string& channel,
-          const base::ListValue& arguments) {
-  RenderFrame* render_frame = GetCurrentRenderFrame();
-  if (render_frame == nullptr)
-    return;
-
-  bool success = render_frame->Send(new AtomFrameHostMsg_Message(
-      render_frame->GetRoutingID(), internal, channel, arguments));
-
-  if (!success)
-    args->ThrowError("Unable to send AtomFrameHostMsg_Message");
-}
-
-base::ListValue SendSync(mate::Arguments* args,
-                         bool internal,
-                         const std::string& channel,
-                         const base::ListValue& arguments) {
-  base::ListValue result;
-
-  RenderFrame* render_frame = GetCurrentRenderFrame();
-  if (render_frame == nullptr)
+class IPCRenderer : public mate::Wrappable<IPCRenderer> {
```

Comment:
@zcbenz it looks like you moved away from this pattern in d16b5811409445724c1db9d5f95b5aa9ad236be8, do you have thoughts on whether it's OK here or will this leak also?

Reply:
This will leak too since instances may be created in subframes and currently we don't cleanup Node environments there.

But this leak should be fine as it only stores a pointer, unlike `SpellChecker` which takes a large chuck of memory.

Reply:
> currently we don't cleanup Node environments there.

FWIW in Electron 6 we may be able to do this, we already do clean them up when you enable `nodeIntegrationSupportInSubFrames`.  Enabling this for all frames should be a thing we can _try_ on master

Reply:
Perhaps this should implement `RenderFrameObserver` and delete itself when the render frame is deleted?

Reply:
@nornagon The issue isn't that it outlives the frame, rather it comes into existence *before* the document has been created inside the frame.